### PR TITLE
Microsite/reusable components

### PR DESCRIPTION
**Pull request by ramytanios**

### CODE REVIEW

File: docs/examples/reusable-components.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        options.map { name =>
+          option(
+            selected := (name === selected0(state)), // use of Eq[O] implicit
+            key := name.show, // use of Show[O] implicit
+            value := name.show,
+            name.show
+          )
+        }
+      )
+    }
```
Comment (buntec) [inline docs/examples/reusable-components.md:63]: We can use pristine argument names and rename the imports instead
```suggestion
  def customSelect[O: Show: Eq](
      fromString: String => Option[O],
      onChange: (S, O) => Option[A],
      options: List[O],
      selected: S => O
  )(implicit dsl: ff4s.Dsl[F, S, A]): dsl.V = {
    import dsl._
    import dsl.html.{onClick => onClick0, selected => selected0, _}
    useState { state =>
      select(
        onChange0 := ((ev: dom.Event) =>
          ev.target match {
            case el: dom.HTMLSelectElement =>
              fromString(el.value).flatMap(onChange(state, _))
            case _ => None
          }
        ),
        options.map { name =>
          option(
            selected0 := (name === selected(state)),
            key := name.show,
            value := name.show,
            name.show
          )
        }
      )
    }
```

File: docs/examples/reusable-components.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    import dsl.html._
+
+    useState { state =>
+      button(
+        child,
+        disabled := isDisabled(state),
+        onClick := (_ => onClick0(state))
+      )
+    }
+  }
```
Comment (buntec) [inline docs/examples/reusable-components.md:81]: Similarly as above
```suggestion
  def customButton(
      dsl: ff4s.Dsl[F, S, A]
  )(child: dsl.V, onClick: S => Option[A], isDisabled: S => Boolean): dsl.V = {
    import dsl._
    import dsl.html.{onClick => onClick0, _}

    useState { state =>
      button(
        child,
        disabled := isDisabled(state),
        onClick0 := (_ => onClick(state))
      )
    }
  }
```

File: docs/examples/reusable-components.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+import cats.syntax.all._
+
+object View {
+
+  def apply[F[_]](implicit dsl: ff4s.Dsl[F, State, Action]) = {
+
+    import dsl._
+    import dsl.html._
+
+    // create a `Components` instance and import components
```
Comment (buntec) [inline docs/examples/reusable-components.md:217]: Obvious comments just add visual noise.
```suggestion
```

File: docs/examples/reusable-components.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  // try to parse a string to a `CurrencyPair`
+  def fromString(s: String): Option[CurrencyPair] = s match {
+    case "EURUSD" => EURUSD.some
+    case "USDCHF" => USDCHF.some
+    case "GBPUSD" => GBPUSD.some
+    case "CHFEUR" => CHFEUR.some
+    case _        => none
+  }
+
+  // all available currency pairs
```
Comment (buntec) [inline docs/examples/reusable-components.md:143]:
```suggestion
```

File: docs/examples/reusable-components.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+sealed trait CurrencyPair
+
+object CurrencyPair {
+  // In this example, we choose few currency pairs
+  case object EURUSD extends CurrencyPair
+  case object USDCHF extends CurrencyPair
+  case object GBPUSD extends CurrencyPair
+  case object CHFEUR extends CurrencyPair
+
+  // try to parse a string to a `CurrencyPair`
```
Comment (buntec) [inline docs/examples/reusable-components.md:134]:
```suggestion
```

File: docs/examples/reusable-components.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+```
+
+In this example, the elements of `customSelect` are currency pairs. A generic currency pair is modelled by a
+`CurrencyPair` trait. Specific pairs are `case object`'s extending from the latter.
+
+```scala mdoc:js:shared
+import cats.Show
+import cats.kernel.Eq
+
+sealed trait CurrencyPair
```
Comment (buntec) [inline docs/examples/reusable-components.md:125]: I know this is a toy example, but I'm not very fond of this encoding. There are ~180 currencies, so this wouldn't scale in practice.

File: docs/examples/reusable-components.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+      div(
+        customSelect[CurrencyPair](
+          CurrencyPair.fromString,
+          (_, pair) => SetCurrencyPair(pair).some,
+          CurrencyPair.all,
+          _.currencyPair
+        ),
+        div(s"Selected currency pair: ${state.currencyPair}"),
+        customButton(dsl)(
+          span("Get FX Rate"),
```
Comment (buntec) [inline docs/examples/reusable-components.md:231]: If we are not using the button in its full generality, then we might as well make it accept a string for a label and get the benefit of passing `dsl` implicitly.

**Review by buntec — CHANGES_REQUESTED** [review]
Overall I find the example a bit too long, so let's try to tighten it up and remove unnecessary fluff 🤓

File: docs/examples/reusable-components.md (hunk 1 above)
Comment (ramytanios) [reply docs/examples/reusable-components.md:63]: nice one!

**Comment by ramytanios** [comment]
thanks for the feedback. 

1. The reason why I am using a button with an explicit `dsl` is to have an example here a dsl is needed explicitly (since for the select it is implicit). I think its nice to have these two examples. And then I use a `span` for the dsl in the view. Of course I could have added something more complex than a `span` but just wanted to keep it simple. 
2. For the currency pairs encoding -- totally agree. However I wanted to show also how the `[O: Show: Eq]` can be nicely used for some objects of type `O` in a select component (while keeping the example in the context of the FX rates API). We can certainly work with strings if you prefer and shrink the size of the example.

Let me know ;)

File: docs/examples/reusable-components.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
+# Reusable Components
+
+Reusable UI components are crucial in frontend development allowing developers to impose style uniformity and avoid code repeatability.
+Button and select components are two of the most popular UI components and will be used in this example.
```
Comment (buntec) [inline docs/examples/reusable-components.md:4]:
```suggestion
The ability to factor out and reuse components, possibly across project boundaries, is an essential feature of any UI framework or library. To accomplish this in ff4s we have to work with generic state and action types. Here we illustrate the general approach with some toy examples.
```

File: docs/examples/reusable-components.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+Components can be implemented as member methods of a class with type parameters `F` for the effect type, `S` for the state type and `A` for the action type.
+
+```scala mdoc:js:shared
+import org.scalajs.dom
+import cats.Show
+import cats.kernel.Eq
+import cats.syntax.all._
+
+class Components[F[_], S, A] {
```
Comment (buntec) [inline docs/examples/reusable-components.md:16]:
```suggestion
// S and A are the state and action types, respectively
trait Elements[F[_], S, A] {
```

File: docs/examples/reusable-components.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
+# Reusable Components
+
+Reusable UI components are crucial in frontend development allowing developers to impose style uniformity and avoid code repeatability.
+Button and select components are two of the most popular UI components and will be used in this example.
+
+## Components
+
+Components can be implemented as member methods of a class with type parameters `F` for the effect type, `S` for the state type and `A` for the action type.
```
Comment (buntec) [inline docs/examples/reusable-components.md:8]:
```suggestion
We recommend organizing components into generic traits:
```

File: docs/examples/reusable-components.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+```
+
+## Store
+
+The construction of the store is straightforward and omitted for brevity.
+
+## View
+
+The interesting bit of the `View` is the instantiation of the `Components` class that allows
+us to import and use all components created.
```
Comment (buntec) [inline docs/examples/reusable-components.md:128]:
```suggestion
With concrete state and action types we can instantiate our components and build the view.
```

File: docs/examples/reusable-components.md [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+If this is not the case, components must require the dsl to be passed explicitly (in this case the `customButton`).
+This is particularly powerful as we can pass complex HTML elements (`div`, `span`, etc ..) to the button component.
+
+## State
+
+```scala mdoc:js:shared
+final case class State(counter: Int = 0, fruit: Fruit = Fruit.Banana)
+```
+
+For this example, elements of the select component are fruits each of type `Fruit` and modelled as follows:
```
Comment (buntec) [inline docs/examples/reusable-components.md:78]:
```suggestion
To show off our custom selection component we define an enumeration with the required `Eq` and `Show` instances.
```

File: docs/examples/reusable-components.md [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+sealed trait Action
+case class SetFruit(fruit: Fruit) extends Action
+case object Inc extends Action
+```
+
+## Store
+
+The construction of the store is straightforward and omitted for brevity.
+
+## View
```
Comment (buntec) [inline docs/examples/reusable-components.md:125]:
```suggestion
```

File: docs/examples/reusable-components.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
+# Reusable Components
+
+The ability to factor out and reuse components, possibly across project boundaries, is an essential feature of any UI framework or library. To accomplish this in ff4s we have to work with generic state and action types. Here we illustrate the general approach with some toy examples.
+
+We recommend organizing components into generic traits:
+
+```scala mdoc:js:shared
+import org.scalajs.dom
+import cats.Show
+import cats.kernel.Eq
```
Comment (buntec) [inline docs/examples/reusable-components.md:10]:
```suggestion
import cats._
```

File: docs/examples/reusable-components.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+import cats.syntax.all._
+
+// S and A are the state and action types, respectively
+trait Elements[F[_], S, A] {
```
Comment (buntec) [inline docs/examples/reusable-components.md:14]:
```suggestion
trait Selects[F[_], S, A] {
```

File: docs/examples/reusable-components.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            key := name.show,
+            value := name.show,
+            name.show
+          )
+        }
+      )
+    }
+
+  }
+
```
Comment (buntec) [inline docs/examples/reusable-components.md:46]:
```suggestion
// more ...

}

trait Buttons[F[_], S, A] {
```

File: docs/examples/reusable-components.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+}
+```
+
+Interesting to note is how an instance of `ff4s.Dsl` is passed as parameter to the components.
+For components whose parameters' types do not depend on the dsl, the latter is passed as an implicit (in this case the `customSelect` component).
+If this is not the case, components must require the dsl to be passed explicitly (in this case the `customButton`).
+This is particularly powerful as we can pass complex HTML elements (`div`, `span`, etc ..) to the button component.
+
+## State
```
Comment (buntec) [inline docs/examples/reusable-components.md:70]:
```suggestion
Unfortunately, if a component takes another component as a parameter, then `dsl` cannot be passed implicitly due to dependent typing.
```

File: docs/examples/reusable-components.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+object View extends {
+
+  def apply[F[_]](implicit dsl: ff4s.Dsl[F, State, Action]) = {
+
+    import dsl._
+    import dsl.html._
+
+    object Components extends Elements[F, State, Action]
+    import Components._
```
Comment (buntec) [inline docs/examples/reusable-components.md:132]:
```suggestion
    object components extends Selects[F, State, Action] with Buttons[F, State, Action]
    import components._
```

File: docs/examples/reusable-components.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+}
+```
+
+```scala mdoc:js:shared
+sealed trait Action
+case class SetFruit(fruit: Fruit) extends Action
+case object Inc extends Action
+```
+
+The construction of the store is straightforward and omitted for brevity.
```
Comment (buntec) [inline docs/examples/reusable-components.md:117]:
```suggestion
We omit the straightforward definition of the store.
```

File: docs/examples/reusable-components.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+
+    useState { state =>
+      button(
+        child,
+        disabled := isDisabled(state),
+        onClick0 := (_ => onClick(state))
+      )
+    }
+  }
+
```
Comment (buntec) [inline docs/examples/reusable-components.md:61]:
```suggestion

// more ...

```

**Review by buntec — CHANGES_REQUESTED** [review]

**Review by buntec — APPROVED** [review]

**Comment by buntec** [comment]
Thanks so much!

**Pull request merged**